src/uschedule.1.in
.TH uschedule 1 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uschedule - uschedule a job
.LINE
.section SYNOPSIS
\fBuschedule\fR [OPTIONS] ID TIMESPEC [...]
.section DESCRIPTION
\fBuschedule\fR schedules the command with the identifier \fIID\fR 
to be executed at the time specified by \fITIMESPEC\fR. Multiple
\fITIMESPEC\fR arguments are allowed.
.paragraph
\fIID\fR is the identifier of a command previously registered
with
.manlink uschedulecmd.html uschedulecmd(1) .
.section OPTIONS
.options
.option "-., --dot-as-home"
The current working directory will be used instead of $HOME.
.option "-1, --null1"
Redirect the standard output of the job to /dev/null. The default is to
write it into the log file of the 
.manlink uscheduled.html uscheduled(8)
daemon.
.option "-2, --null2"
Redirect the standard error output of the job to /dev/null. The same
default applies.
.option "-c, --count=NNN"
Repeat the command up to \fINNN\fR times. A value of \fI1\fR means the
the job will run once, \fI0\fR is a synonym for \fIforever\fR, which
is the default.
.option "-d, --dir=DIR"
Put the new job into \fIDIR\fR. The default is \fI~/.uschedule/\fR.
.option "-D, --description=DESC"
Give the new job the description \fIDESC\fR. The description is used
for the user interface commands 
.manlink uschedulelist.html uschedulelist(1)
and
.manlink uschedulerm.html uschedulerm(1)
only.  Descriptions must not be longer than 70 characters and must
not contain colons.
.option "-e, --every=NUMBER[mhdw]"
Repeat every \fINUMBER\fR time units (default: seconds).

This option is implemented in such a way that the NUMBER is added once 
at the start of a search. Then all other restrictions (late, from, to, 
TIMESPEC) will be applied and the next matching time will be searched 
for. In other words: The start time to be searched from is changed from
\fBnow\fR to \fBnow + NUMBER\fR.

A \fIm\fR (\fIh\fR, \fId\fR or \fIw\fR) appended to \fINUMBER\fR changes 
the time unit from seconds to minutes (hours, days or weeks).
.option "-f, --from=TIMESPEC"
Jobs will only be started if \fITIMESPEC\fR has matched. This
together with the \fI--to\fR option allows to further restrict the
times a job will be started. This example starts a job JOBID every day
a few seconds after midnight, but only after the first of april, 2004,
has been reached:
.paragraph
.nf
.HTML <pre>
   uschedule --from "2003-4-1 00:00:00" JOBID '*-*-* 00:00:10'
.HTML </pre>
.fi
.optionparagraph
TIMESPEC should be simple. Wild cards are only allowed at the start,
\fInot\fR after any fixed number. \fI2002-*-1 00:00:00\fR is forbidden,
\fI*-*-* *:00:00\fR is OK.
.optionparagraph
Weekday names may be used, too, though the result is likely to be 
non-intuitive. Better avoid them.
.optionparagraph
This option was added in version 0.6.0.
.option "-l, --late=SECONDS"
Allow the job to be executed up to \fISECONDS\fR late. This is useful
if the machine or the 
.manlink uscheduled.html uscheduled(8)
daemon was down during the time the job should have run.
.optionparagraph
The default is 3600 seconds (one hour).
.option "-t, --to=TIMESPEC"
Jobs will only be started if \fITIMESPEC\fR has not been reached. This
together with the \fI--from\fR option allows to further restrict the
times a job will be started. The example below starts a job every day
a few seconds after midnight, but only up to 30th of march, 2004:
.paragraph
.nf
.HTML <pre>
   uschedule --to "2003-4-1 00:00:00" jobid '*-*-* 00:00:10'
.HTML </pre>
.fi
.optionparagraph
TIMESPEC should be simple. Wild cards are only allowed at the start,
\fInot\fR after any fixed number. \fI2002-*-1 00:00:00\fR is forbidden,
\fI*-*-* *:00:00\fR is OK.
If wild cards are used, the --from option has to be used, to.
.optionparagraph
Weekday names may be used, too, though the result is likely to be 
non-intuitive. Better avoid them.
.optionparagraph
This option was added in version 0.6.0.
.section "TIMESPEC"
A time specification consists of two or three words. The first, optional,
words specifies a day-of-week, the next the year, month and day-of-month,
and the last word specifies hour, minute and seconds.
.br
Words are separated by exactly one space.
.subsection "Day Of Week"
The day of the week is given as a comma separated list of weekday
names or three letter abbreviations thereof. Names are case insensitive.
The default is to run the job at any day of the week.
.paragraph
\fISunday,Wed\fR is a valid list. \fIMonday, Tues\fI isn't.
.subsection "Date"
The date consists of three parts: Year, month and day. Two parts are separated
by a single dash. Each part is a numerical value as described below.
.subsection "Time"
The time consists of three parts, too. Hour, minute and second are
by a single colon. Each part is a numerical value as described below.
.subsection "Number Specification"
Whenever is number is allowed in a job execution time specification
is may be either a single value, a single star ("*", meaning "all
possible values") or a comma 
separated list of values.
.br
A value may be a number or a number ("a") followed by a slash 
followed by a number ("b"), meaning "execute at 'a+b*n'", where
"n" is a positive integer including 0). You may use a plus sign
instead of the slash ('a+b').
.br
Note that 30/10 (oder 30+10) and such things are understood quite literally,
meaning "30, 40, 50" and not "0, 10, 20, 30, 40, 50".

.subsection "Examples"
The following schedules a job to be run at midnight of each seventh
of the month:
.nf
.HTML <pre>
   *-*-7 00:00:00
.HTML </pre>
.fi
.paragraph
To run a job on every monday in december at 12:00:00:
.paragraph
.nf
.HTML <pre>
   Monday *-12-* 12:00:00
.HTML </pre>
.fi
.paragraph
To run a job 30 minutes and 45 seconds after each full hour on every
monday and friday if that day is the first or third day of the
month in the months january, march, may, juli, september and 
november:
.paragraph
.nf
.HTML <pre>
   mon,fri *-1/2-1,3 *:30:45
.HTML </pre>
.fi
.paragraph
.subsection "Incomplete time specs"
\fIuschedule\fR attempts to complete the TIMESPECs, except for the 
--from or --to ones. \fI*-\fR is used when year or month is missing,
\fI*\fR is used when the day is missing, and
\fI*:\fR is used when hour or minute are missing. If there is no
time given at all, then \fI0:0:0\fR is used. Examples:
.paragraph
.nf
.HTML <pre>
  03-05 08:05:40      -> *-03-05 08:05:40
  05 08:05:40         -> *-*-05 08:05:40
  08:05:40            -> *-*-* 08:05:40
  05:40               -> *-*-* *:05:40
  40                  -> *-*-* *:*:40
  Sat,Sun 05 08:05:40 -> Sat,Sun *-*-05 08:05:40
  Sat,Sun 08:05:40    -> Sat,Sun *-*-* 08:05:40
  2003-03-05 05:40    -> 2003-03-05 *:05:40
  2003-03-05          -> 2003-03-05 0:0:0
  03-05               -> *-03-05 0:0:0
.HTML </pre>
.fi
.paragraph
.section "TIME OFFSET"
An argument consisting of a plus sign and one to four numbers separated 
by colons (+[[[dd:]hh:]mm:]ss)
means to start the job once in the future, at the time reached with the
current time is added to the argument. dd is the offset in days, hh
in hours, mm in minutes und ss stands for seconds. That is, 
.nf
.HTML <pre>
  uschedule test +1:0:0:0 
  uschedule test +24:0:0
  uschedule test +1440:0
  uschedule test +86400
.HTML </pre>
.fi
all start the job "test" exactly one day after uschedule is executed.
.paragraph
.section DIFFERENCES
The unix \fIcron\fR daemon executes a job if either day or week 
or day of month matches. 
.manlink uscheduled.html uscheduled(8)
executes it if \fIboth\fR match. The is no way to completely mimic either 
logic with the other software.
.paragraph
Unix \fIcron\fR often needs a separated \fIat\fR daemon to execute 
one-time-jobs. This is nothing more than a design problem in cron.
.paragraph
Unix also provides a \fIbatch\fR command, which executes jobs if
the system load is low. This is often implemented inside the \fIat\fR
daemon. \fIuschedule\fR doesn't provide a way to emulate this (note
that \fIbatch\fR is pointless anyway: if you just have a short job
then you might as well run it, it doesn't matter. If you have a 
long and resource-intensive job and don't want your job to 
eat resources when there's a shortage of them then \fIbatch\fR 
provides no way to temporarily suspend a job during this time.
If the resource in question is not system load but memory,
disk- or network-bandwidth then \fIbatch\fR doesn't help you).
.paragraph
Unix \fIcron\fR doesn't support second granularity. \fIuschedule\fR
does this for only one reason: to help to avoid that all jobs of all
users start in the same second and overload the system.
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedulecmd.html uschedulecmd(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uschedule.c
#include "uogetopt.h"
#include "scheduled.h"
#include "bailout.h"
#include "str.h"
#include "error.h"
#include "scan.h"
#include "fmt.h"
#include <unistd.h>

static unsigned long o_repeats; 
static unsigned long o_late=3600;
static unsigned long o_every=0;
static const char *o_every_s;
static const char *o_comment;
static const char *o_dir;
static char *o_from;
static char *o_to;
static int o_verbose=0;
static int o_null1=0;
static int o_null2=0;
static int o_dot_as_home;
static uogetopt2 myopts[]={
{'.',"dot-as-home",uogo_flag,UOGO_NOARG,&o_dot_as_home, 1 ,
"Use current directory instead of $HOME.",0,0},
{'1',"null1",uogo_flag,UOGO_NOARG,&o_null1, 1 ,
"Redirect stdout to /dev/null.",
"Redirect the standard output of the job to "
"/dev/null. The default is to print it into the "
"same log the scheduled daemon uses.",0},
{'2',"null2",uogo_flag,UOGO_NOARG,&o_null2, 1 ,
"Redirect stderr to /dev/null.",
"Redirect the standard error output of the job to "
"/dev/null. The default is to print it into the "
"same log the scheduled daemon uses.",0},
{'c',"count",uogo_ulong,0,&o_repeats, 0 ,
"Repeat job N times.",
"The jobs will be deleted after it was executed "
"or or tried to be executed the given number of "
"times. The default is to repeat it forever.","N"},
{'d',"dir",uogo_string,0,&o_dir, 0 ,
"Jobdirectory.",
"Jobs will be written to this directory.","DIR"},
{'D',"description",uogo_string,0,&o_comment, 0 ,
"Short job description.",
"Less than 70 characters, no colons.", "STRING"},
{'e',"every",uogo_string,0,&o_every_s, 0 ,
"Repeat every NUMBER time units (default: seconds).",
"To change the time unit append these characters to the NUMBER: "
"`m' (minutes), `h' hours, `d' days and `w' weeks.\n"
"This option is implemented in such a way that the NUMBER is added once "
"at the start of a search. Then all other restrictions (late, from, to, "
"TIMESPEC) will be applied and the next matching time will be searched "
"for.", "NUMBER"},
{'f',"from",uogo_string,0,&o_from, 0 ,
"Run job only after STARTSPEC.",
"STARTSPEC is a TIMESPEC. Jobs will only be started "
"after STARTSPEC is reached. This may be used "
"together with --to.\n"
"Note: leading wildcards are OK, trailing wildcards "
"are not! See the manual page.", "STARTSPEC"},
{'l',"late",uogo_ulong,0,&o_late, 0 ,
"Allow job to start up to SECONDS later.",
"The default is to allow it to be executed up to one hour later.","SECONDS"},
{'t',"to",uogo_string,0,&o_to, 0 ,
"Run job only before ENDSPEC.",
"ENDSPEC is a TIMESPEC. Jobs will only be started "
"until ENDSPEC is reached. This may be used together "
"with --from.\n"
"Note: leading wildcards are OK, trailing wildcards "
"are not! See the manual page.", "ENDSPEC"},
{'v',"verbose",uogo_flag,UOGO_NOARG,&o_verbose, 1 ,
"Print success messages.",0,0},
{0,0,0,0,0,0,0,0,0}
};              

static void
create_job(struct jobinfo *j)
{
	static stralloc fn;
	static stralloc fn2;

	make_name(&fn2,j);

	if (!stralloc_copys(&fn,IDDIR)) oom();
	if (!stralloc_cats(&fn,"/")) oom();
	if (!stralloc_catb(&fn,j->id,j->idlen)) oom();
	if (!stralloc_0(&fn)) oom();
	if (-1==link(fn.s,fn2.s))
		xbailout(111,errno,"failed to link ",fn.s, " to ", fn2.s);
}
/* the matching alghorithm for the from/to specifications does 
 * very unintuitive things with tailing wildcards. That is, 
 * a wildcard follows a fixed value. */
static int 
wildcardcheck(stralloc *sa,const char *name)
{
  unsigned int i;
  unsigned int bits=0;
  unsigned int expect=0;
  for (i=0;i<sa->len;i++) {
    switch(sa->s[i]) {
    case 'Y': bits|=32; expect=63; break;
    case 'M': bits|=16; if (!expect) expect=31; break;
    case 'D': bits|=8; if (!expect) expect=15; break;
    case 'h': bits|=4; if (!expect) expect=7; break;
    case 'm': bits|=2; if (!expect) expect=3; break;
    case 's': bits|=1; if (!expect) expect=1; break;
    case 'W': 
      warning(0,"Weekdays in ",name," specifications are unlikely to "
	"work as you expect.",0);
    }
  }
  if (expect!=bits)
    xbailout(2,0,name," specification contains wildcards after "
      "fixed values.",0,0);
  return bits;
}
static uogetopt_env myoptenv={
"uschedule",PACKAGE,VERSION,
"uschedule [options] ID TIMESPEC [...]",
"This program schedules the command ID to be run at TIMESPEC.",
"long",
"Report bugs to uschedule@lists.ohse.de",
3,0,0,0,uogetopt_out,myopts
};

int main(int argc, char **argv) 
{
	const char *id;
	unsigned int i;
	stralloc sa_from=STRALLOC_INIT;
	stralloc sa_to=STRALLOC_INIT;
	bailout_progname(argv[0]);
	flag_bailout_fatal_begin=3;
	uogo_posixmode=1;
	myoptenv.program=flag_bailout_log_name;
	uogetopt_parse(&myoptenv,&argc,argv);
	if (o_every_s) {
	  unsigned int x;
	  x=scan_ulong(o_every_s,&o_every);
	  switch (o_every_s[x]) {
	  case 's': o_every*=1; break; /* be friendly */
	  case 'm': o_every*=60; break;
	  case 'h': o_every*=360; break;
	  case 'd': o_every*=86400; break;
	  case 'w': o_every*=86400*7; break;
	  case 0: break;
	  default:
	    xbailout(100,0,"failed to parse --every argument: ",o_every_s,0,0);
	  }
	}

	if (o_comment)
		check_id(o_comment);
	if (o_to) {
		int bits;
		parse_timespec(&sa_to,o_to);
		bits=wildcardcheck(&sa_to,"to");
		if (bits!=63 && !o_from)
		  xbailout(2,0,"--to specifications with wildcards "
		                "need a --from specification, too",0,0,0);
	}
	if (o_from)  {
		parse_timespec(&sa_from,o_from);
		wildcardcheck(&sa_from,"from");
	}

	change_dir(0,o_dir,o_dot_as_home);
	id=argv[1];
	for (i=2;argv[i];i++) {
		struct jobinfo j;
		static stralloc sa;
		static stralloc sa2;

		HACK_TAIA_SEC(&j.lastrun)=0;
		j.lastrun.nano=0;
		j.lastrun.atto=0;
		j.late=o_late;
		j.id=id;
		j.idlen=str_len(id);
		j.cronspec=0; /* dito */
		j.cronlen=0; /* dito */
		j.null1=o_null1;
		j.null2=o_null1;
		j.comment=o_comment;
		if (o_comment)
			j.commentlen=str_len(o_comment);
		else
			j.commentlen=0;
		if ('+'==argv[i][0]) {
		  timespec_from_now(&sa,argv[i]+1);
		  j.repeats=1;
		} else {
		  fill_timespec(&sa2,argv[i]); /* incomplete -> complete */
		  parse_timespec(&sa,sa2.s);
		  j.repeats=o_repeats;
		}
		j.cronspec=sa.s;
		j.cronlen=sa.len;
		/* the following work since the sa_* are initialized to 0 */
		j.fromlen=sa_from.len; 
		j.fromspec=sa_from.s;
		j.tolen=sa_to.len; 
		j.tospec=sa_to.s;
		j.every=o_every;
		  
		create_job(&j);
	}
	notice();
	return 0;
}
